### PR TITLE
Use libgc for all malloc calls, so they are scanned properly.

### PR DESCRIPTION
**Pull request by ChrisDodd**
- glibstdc++ uses malloc to allocate exceptions being thrown, which
  could result in them not being scanned by the garbage collector,
  leading to memory corruption.

### CODE REVIEW

File: lib/gc.cpp [hunk 1]
```diff
@@ -33,26 +34,57 @@ limitations under the License.
...
      * memory, we need to force initialization */
     if (!done_init) {
+        started_init = true;
         GC_INIT();
         done_init = true; }
     return ::operator new(size, UseGC, 0, 0);
 }
 void *operator new[](std::size_t size) {
     if (!done_init) {
+        started_init = true;
```
Comment (cc10512) [inline lib/gc.cpp]: why not use `std::call_once`?

File: lib/gc.cpp [hunk 2]
```diff
@@ -33,26 +34,57 @@ limitations under the License.
...
         GC_INIT();
         done_init = true; }
     return ::operator new(size, UseGC, 0, 0);
 }
 void operator delete(void *p) _GLIBCXX_USE_NOEXCEPT { return gc::operator delete(p); }
 void operator delete[](void *p) _GLIBCXX_USE_NOEXCEPT { return gc::operator delete(p); }
 
+void *realloc(void *ptr, size_t size) {
```
Comment (cc10512) [inline lib/gc.cpp]: Why aren't you using the glibc hooks? https://www.gnu.org/software/libc/manual/html_node/Hooks-for-Malloc.html

File: lib/gc.cpp (hunk 1 above)
Comment (ChrisDodd) [reply lib/gc.cpp]: `call_once` doesn't work in the recursive call case -- we need to detect if we've started to do GC_INIT but it has not yet finished.

File: lib/gc.cpp (hunk 2 above)
Comment (ChrisDodd) [reply lib/gc.cpp]: Do they work on OSX libc?

**Pull request merged**